### PR TITLE
Bug fix and improvements for [z]imatcopy interface.

### PR DESCRIPTION
**Pull request by klho**
Fix/implementation for #4038.

### CODE REVIEW

**Comment by brada4** [comment]
Sorry

**Comment by martin-frbg** [comment]
> Why do you introduce memory leak claiming bug fix and memory savings?

@brada4 Why do you have to pop up here every now and again like a playground bully ? You have been warned several times that this behavior is not acceptable. If you cannot state your _technical_ objections in a civil manner, please refrain from alienating contributors here. Replying in your native language is perfectly fine if it is a language problem for you.

File: interface/zimatcopy.c [hunk 1]
```diff
@@ -125,95 +125,93 @@ void CNAME( enum CBLAS_ORDER CORDER, enum CBLAS_TRANSPOSE CTRANS, blasint crows,
...
 
 	if ( order == BlasColMajor )
 	{
 
 		if ( trans == BlasNoTrans )
 		{
-	  		OMATCOPY_K_CN(*rows, *cols, alpha[0], alpha[1], a, *lda, b, *ldb );
-	  		OMATCOPY_K_CN(*rows, *cols, (FLOAT) 1.0, (FLOAT) 0.0 , b, *ldb, a, *ldb );
-			free(b);
-			return;
```
Comment (mmuetzel) [inline interface/zimatcopy.c:195]: The early `return` here might have been a micro-optimization.
Would it make sense to make this an `else if` ladder when removing the early `return` to safe a couple of comparison operations?

Comment (klho) [reply interface/zimatcopy.c:195]: Agreed, thanks -- will make the changes.

File: interface/zimatcopy.c [hunk 2]
```diff
@@ -125,95 +125,93 @@ void CNAME( enum CBLAS_ORDER CORDER, enum CBLAS_TRANSPOSE CTRANS, blasint crows,
...
-        {
-                printf("Memory alloc failed in zimatcopy\n");
-                exit(1);
-        }
+	msize = (size_t)(*rows) * (*cols) * sizeof(FLOAT) * 2;
 
+	b = malloc(msize);
+	if ( b == NULL )
+	{
+		printf("Memory alloc failed\n");
```
Comment (mmuetzel) [inline interface/zimatcopy.c]: Should the original (more detailed) message be kept here?

Comment (klho) [reply interface/zimatcopy.c]: Sorry, I just copied this over from imatcopy.c (for consistency of tabs/spaces) and failed to notice that the message is materially different. I'll change it back to the more detailed message -- and also add the same more detailed print out to imatcopy.c as well.

File: interface/imatcopy.c [hunk 3]
```diff
@@ -120,71 +120,71 @@ void CNAME( enum CBLAS_ORDER CORDER, enum CBLAS_TRANSPOSE CTRANS, blasint crows,
...
-        return; 
     }
-
 #endif
 
-	if ( *lda >  *ldb )
-		msize = (size_t)(*lda) * (*ldb)  * sizeof(FLOAT);
-	else
-		msize = (size_t)(*ldb) * (*ldb)  * sizeof(FLOAT);
+	msize = (size_t)(*rows) * (*cols) * sizeof(FLOAT);
```
Comment (mmuetzel) [inline interface/imatcopy.c:155]: ~~I'm not sure about the background for the different permutations of the product here. Maybe, it was for numeric stability/overflow behavior/...?
Are you sure this is no longer needed? (Removing it seems unrelated to the other changes. But I could be wrong.)~~

Sorry. Misread the existing code.

Comment (mmuetzel) [reply interface/imatcopy.c:155]: Also, what if `*ldb` is larger than the corresponding `*rows` or `*cols`? Wouldn't the allocated memory be too small with the new logic?

File: interface/zimatcopy.c (hunk 1 above)
Comment (martin-frbg) [reply interface/zimatcopy.c:195]: Thanks

File: interface/imatcopy.c (hunk 3 above)
Comment (martin-frbg) [reply interface/imatcopy.c:155]: My reading of the MKL documentation (in lieu of an actual standard, given that it is a BLAS extension) also suggests that  either of `*lda,*ldb` may be greater than `*rows` or `*cols`. (I fear I may need to sketch a few cases on paper to see how this works out in practice though... but seeing that this survived in the codebase for almost a decade, the original assumptions must have been "mostly" correct )

Comment (mmuetzel) [reply interface/imatcopy.c:155]: IIUC, the original code over-allocates memory if `*lda` is larger than the actual number of `*rows` or `*cols` it corresponds to.
The proposed change under-allocates if `*ldb` is larger than the actual number of `*rows` or `*cols` it corresponds to.

Edit: No. It's more complicated than that with the possible transposition.
Only the second part about the proposed code potentially under-allocating memory is true. ...I think...

Comment (mmuetzel) [reply interface/imatcopy.c:155]: The following could be correct:
```
if ( (trans == BlasNoTrans) != (order == BlasColMajor) )
  // output is "input row" major
  msize = (size_t)(*rows) * (*ldb) * sizeof(FLOAT);
else
  // output is "input column" major
  msize = (size_t)(*cols) * (*ldb) * sizeof(FLOAT);
```

But please double check.

Comment (mmuetzel) [reply interface/imatcopy.c:155]: Updated that snippet twice forth and back. That's a bigger head-scratcher than I'd like to admit.
But I think the last iteration is correct now. (As was the first. But I didn't believe in myself. 😉)

Comment (klho) [reply interface/imatcopy.c:155]: I find it most helpful to think about how many entries are copied from `a` to `b` and then back again to `a`: the answer is simply `rows * cols`, regardless of the input (`lda`) and output (`ldb`) strides. (As a motivating example, consider transposing just a `10 x 10` block of a `10000 x 10000` matrix.)

Note, importantly, that `ldb` doesn't really have anything to do with `b` (maybe this is the cause of the confusion): `ldb` is the stride for the output matrix, which is stored in `a`, specifying where/how to write the data coming from `b`; while `b` just needs to hold that data, in principle with any stride it likes (as a purely internal implementation detail) -- but ideally stored as compactly as possible. From this point of view, it's clear that `msize` (= size of `b`) should not depend on `ldb` at all. (A more suggestive renaming of variables might read: `a` -> `ab` (as in the MKL documentation), `b` -> `tmp`, and `msize` -> `tmpsize`.)

The old code breaks precisely when `trans = BlasNoTrans` (or equivalent) because then `lda` and `ldb` both only give you information on the same dimension (either rows or columns, but not both), as in the opening test case in #4038. (It can be "fixed" by allocating `lda * ldb` entries instead, but this would not be optimal since `lda * ldb >= rows * cols`.)

Comment (mmuetzel) [reply interface/imatcopy.c:155]: I might be misreading the MKL documentation. But afaict, `*ldb` is the stride that the function caller selects for the output matrix. `msize` is the memory in bytes that is allocated for that output.
As an example, assume a column-major output matrix `M` with 2×3 elements and a user-selected stride of 10. In that case, `M[10]` should hold the first value in the second column. If the memory allocated for `M` is only for 2*3=6 elements, that would be out of bounds.

If this is correct, please check if it is really possible to replace `ldb` in the function calls further down in this file.

Comment (mmuetzel) [reply interface/imatcopy.c:155]: Oops. I was reading a page that was a bit misleading (I don't remember where I found it). The documentation here is much clearer:
https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2023-0/imatcopy.html

You are completely right. `a` is the input *and* output buffer. `b` is just a temporary intermediate. Having it exactly match `*rows`*`*cols` is large enough.
Your proposed code looks good to me.

I'm sorry for having caused that confusion.
